### ISSUE TITLE
Nightmare hangs on vo error.

### ISSUE THREAD

**Issue by garibay** [segmentio/nightmare#1303]
I created a simple proxy testing script that's supposed to verify that a proxy works against a particular URL.  If the connection isn't successful within 10 seconds, it's supposed to time out and return a response indicating that it failed.  It works sometimes, but at other times it hangs.  It doesn't seem to make a difference whether it's an http proxy or a socks proxy, or whether or not basic http authentication is being used.

This is the error I see when running under verbose debugging (`DEBUG=*`):

```
C:\var\www\nightmarejs\node_modules\vo\lib\wrap.js:120
    if (str.length > limit) {

TypeError: Cannot read property 'length' of undefined
    at trim (C:\var\www\nightmarejs\node_modules\vo\lib\wrap.js:120:12)
    at C:\var\www\nightmarejs\node_modules\vo\lib\wrap.js:109:14
    at Array.map (native)
    at pretty (C:\var\www\nightmarejs\node_modules\vo\lib\wrap.js:108:30)
    at next (C:\var\www\nightmarejs\node_modules\vo\lib\wrap.js:36:30)
    at Immediate.<anonymous> (C:\var\www\nightmarejs\node_modules\vo\node_modules\co\index.js:52:14)
    at runCallback (timers.js:672:20)
    at tryOnImmediate (timers.js:645:5)
    at processImmediate [as _immediateCallback] (timers.js:617:5)
```

When I was running Node with `DEBUG=nightmare*`, the script seemed to hang after an empty `did-start-loading` response.  But with verbose debugging, I saw that more was happening.  After the error (above), the script does this, indefinitely:

```
Sat, 21 Oct 2017 16:26:29 GMT engine:ws received "2"
Sat, 21 Oct 2017 16:26:29 GMT engine:socket packet
Sat, 21 Oct 2017 16:26:29 GMT engine:socket got ping
Sat, 21 Oct 2017 16:26:29 GMT engine:socket sending packet "pong" (undefined)
Sat, 21 Oct 2017 16:26:29 GMT engine:socket flushing buffer to transport
Sat, 21 Oct 2017 16:26:29 GMT engine:ws writing "3"
Sat, 21 Oct 2017 16:26:54 GMT engine:ws received "2"
Sat, 21 Oct 2017 16:26:54 GMT engine:socket packet
Sat, 21 Oct 2017 16:26:54 GMT engine:socket got ping
Sat, 21 Oct 2017 16:26:54 GMT engine:socket sending packet "pong" (undefined)
Sat, 21 Oct 2017 16:26:54 GMT engine:socket flushing buffer to transport
Sat, 21 Oct 2017 16:26:54 GMT engine:ws writing "3"
Sat, 21 Oct 2017 16:27:19 GMT engine:ws received "2"
Sat, 21 Oct 2017 16:27:19 GMT engine:socket packet
Sat, 21 Oct 2017 16:27:19 GMT engine:socket got ping
Sat, 21 Oct 2017 16:27:19 GMT engine:socket sending packet "pong" (undefined)
Sat, 21 Oct 2017 16:27:19 GMT engine:socket flushing buffer to transport
Sat, 21 Oct 2017 16:27:19 GMT engine:ws writing "3"
...
```

Here is the script:

```
'use strict';

let Nightmare   = require('nightmare');
let vo          = require('vo');
let DateTime    = new Date().toISOString().substr(0, 19).replace(/[^0-9]/g,'');
let useragent   = 'Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/38.0.2125.111 Safari/537.36';

const debug     = (() => {
    if (typeof process.env.DEBUG !== 'undefined' ) {
        return process.env.DEBUG.substr(0, 9) === 'nightmare';
    } else {
        return false;
    }
})();

let host        = (() => { if (process.argv[2] !== null && typeof process.argv[2] !== 'undefined' && process.argv[2] !== '') return process.argv[2]; })();
let port        = (() => { if (process.argv[3] !== null && typeof process.argv[3] !== 'undefined' && process.argv[3] !== '') return process.argv[3]; })();
let user        = (() => {
    if (process.argv[4] !== null && typeof process.argv[4] !== 'undefined' && process.argv[4] !== '') {
        return process.argv[4];
    } else {
        return false;
    }
})();
let pass        = (() => {
    if (process.argv[5] !== null && typeof process.argv[5] !== 'undefined' && process.argv[5] !== '') {
        return process.argv[5];
    } else {
        return false;
    }
})();
let proxy_type  = (() => {
    if (process.argv[6] !== null && typeof process.argv[6] !== 'undefined' && process.argv[6] !== '') {
        return process.argv[6];
    } else {
        return 'http';
    }
})();
let target      = (() => {
    if (process.argv[7] !== null && typeof process.argv[7] !== 'undefined' && process.argv[7] !== '') {
        return process.argv[7];
    } else {
        return 'http://www.ticketmaster.com';//Default site to check if not provided (third argument)
    }
})();
let showWindow  = (() => {
    return (typeof process.argv[8] !== 'undefined' && process.argv[8] !== '');
})();
let settings    = {
    show: showWindow,
    alwaysOnTop: false,
    gotoTimeout: 10000,
    switches: {
        'proxy-server': `${host}:${port}`,
        'ignore-certificate-errors': true
    }
};
let width       = 1366;
let heigth      = 768;

if (proxy_type === 'socks4' || proxy_type === 'socks5') {
    settings.switches.proxyType = proxy_type;
}

if ( ! host || ! port ) process.exit(1);

if (debug) console.log(`arg2: ${process.argv[2]}; arg3: ${process.argv[3]}; arg4: ${process.argv[4]}; arg5: ${process.argv[5]}; arg6: ${process.argv[6]}; arg7: ${process.argv[7]}; arg8: ${process.argv[8]}`);

const expected_title = (() => {
    if (target === 'http://www.example.com') return 'Example Title';
    return '';
})();

function sendMessage(data) {
    if (typeof process.send !== 'undefined') {
        process.send(data);
    } else {
        console.log(`${JSON.stringify(data)}`);
    }
}

vo(function * () {
    let nightmare   = Nightmare(settings);
    let message     = 'passed';
    let title             = '';

    if (user && pass) {
        title = yield nightmare
            .viewport(width, heigth)
            .useragent(useragent)
            .authentication(user, pass)
            .goto(target)//Load page using basic http auth
            .title();
    } else {
        title = yield nightmare
            .viewport(width, heigth)
            .useragent(useragent)
            .goto(target)//Load page
            .title();
    }

    if (title !== expected_title) {
        if (debug) console.log(`unexpected 'title': ${title}`);
        message = 'failed';
    }

    sendMessage({'message': `${message}`});

    yield nightmare.end();
    process.exit();
})(error => {
    if (error) {
        sendMessage({ 'message': 'error', 'status': 'final', 'data': error });
        process.exit();
    } else {
        console.log('done');
    }
});
```

**Comment by garibay** [comment]
I tried modifying the script to not use `vo`, but I still see the same behavior (just without the `TypeError` error message):

Here's the changed part of the script (in place of the `vo` call):

```
if (user && pass) {
    nightmare
        .viewport(width, heigth)
        .useragent(useragent)
        .authentication(user, pass)
        .goto(target)//Load page using basic http auth
        .title()
        .end()
        .then((title) => {
            if (title !== expected_title) {
                if (debug) console.log(`unexpected 'title': ${title}`);
                message = 'failed';
            }

            sendMessage({'message': `${message}`});
        })
        .catch((error) => {
            if (debug) console.log(`error:`, error);

            sendMessage({ 'message': 'error', 'status': 'final', 'data': error });
        });
} else {
    nightmare
        .viewport(width, heigth)
        .useragent(useragent)
        .goto(target)//Load page
        .title()
        .end()
        .then((title) => {
            if (title !== expected_title) {
                if (debug) console.log(`unexpected 'title': ${title}`);
                message = 'failed';
            }

            sendMessage({'message': `${message}`});
        })
        .catch((error) => {
            if (debug) console.log(`error:`, error);

            sendMessage({ 'message': 'error', 'status': 'final', 'data': error });
        });
}
```

And this is what I see in the console output:

```
Sat, 21 Oct 2017 16:47:20 GMT nightmare:log did-get-response-details [{},false,"http://example.net/en-us/css/c784a098c14a4ffd4bf849f8df981299/rebrand_smvp/header_blue.css","http://example.net/en-us/css/c784a098c14a4ffd4bf849f8df981299/rebrand_smvp/header_blue.css",200,"GET","http://www.example.com/",{"accept-ranges":["bytes"],"cache-control":["public, max-age=1657033"],"connection":["close"],"content-encoding":["gzip"],"content-length":["2008"],"content-type":["text/css"],"date":["Sat, 21 Oct 2017 16:47:22 GMT"],"etag":["\"1b09-55749829b53c0\""],"expires":["Thu, 09 Nov 2017 21:04:35 GMT"],"last-modified":["Mon, 21 Aug 2017 20:41:59 GMT"],"server":["Apache"],"vary":["Accept-Encoding"]},"stylesheet"]
Sat, 21 Oct 2017 16:47:21 GMT engine:ws received "2"
Sat, 21 Oct 2017 16:47:21 GMT engine:socket packet
Sat, 21 Oct 2017 16:47:21 GMT engine:socket got ping
Sat, 21 Oct 2017 16:47:21 GMT engine:socket sending packet "pong" (undefined)
Sat, 21 Oct 2017 16:47:21 GMT engine:socket flushing buffer to transport
Sat, 21 Oct 2017 16:47:21 GMT engine:ws writing "3"
Sat, 21 Oct 2017 16:47:46 GMT engine:ws received "2"
Sat, 21 Oct 2017 16:47:46 GMT engine:socket packet
Sat, 21 Oct 2017 16:47:46 GMT engine:socket got ping
Sat, 21 Oct 2017 16:47:46 GMT engine:socket sending packet "pong" (undefined)
Sat, 21 Oct 2017 16:47:46 GMT engine:socket flushing buffer to transport
Sat, 21 Oct 2017 16:47:46 GMT engine:ws writing "3"

```

**Comment by garibay** [comment]
Okay, let's file this one under 'premature reporting' (I swear this has never happened to me before).

I found the problem, and it has nothing to do with the Nightmare script.  The script that calls the Nightmare script (via a `fork` call) uses a timeout so that if the forked child process is still running after 30 seconds, it gets killed.  There was a typo in one of the commands related to that:

`if (test_queue.indexOf(test[id].pid !== -1)) test_queue.splice(index, 1);`

I corrected that:

`if (test_queue.indexOf(test[id].pid) !== -1) test_queue.splice(index, 1);`

And now everything is working correctly...no more "hanging" of the Nightmare script.
